File: tutorials/tour-of-restate-go/part3/ticketobject.go
package main

import (
	"errors"

	"github.com/restatedev/examples/tutorials/tour-of-restate-go/auxiliary"
	restate "github.com/restatedev/sdk-go"
)

type TicketObject struct{}

// <start_reserve>
func (TicketObject) Reserve(ctx restate.ObjectContext, _ restate.Void) (bool, error) {
	status, err := restate.GetAs[auxiliary.TicketStatus](ctx, "status")
	if err != nil && !errors.Is(err, restate.ErrKeyNotFound) {
		return false, err
	}

	if status == auxiliary.TicketStatusAvailable {
		ctx.Set("status", auxiliary.TicketStatusReserved)
		return true, nil
	} else {
		return false, nil
	}
}

// <end_reserve>

// <start_unreserve>
func (TicketObject) Unreserve(ctx restate.ObjectContext, _ restate.Void) (restate.Void, error) {
	status, err := restate.GetAs[auxiliary.TicketStatus](ctx, "status")
	if err != nil && !errors.Is(err, restate.ErrKeyNotFound) {
		return restate.Void{}, err
	}

	if status != auxiliary.TicketStatusSold {
		ctx.Clear("status")
	}

	return restate.Void{}, nil
}

// <end_unreserve>

// <start_mark_as_sold>
func (TicketObject) MarkAsSold(ctx restate.ObjectContext, _ restate.Void) (restate.Void, error) {
	status, err := restate.GetAs[auxiliary.TicketStatus](ctx, "status")
	if err != nil && !errors.Is(err, restate.ErrKeyNotFound) {
		return restate.Void{}, err
	}

	if status == auxiliary.TicketStatusReserved {
		ctx.Set("status", auxiliary.TicketStatusSold)
	}

	return restate.Void{}, nil
}

// <end_mark_as_sold>

File: rust/tutorials/tour-of-restate-rust/src/part4/ticket_object.rs
use crate::auxiliary::ticket_status::TicketStatus;
use restate_sdk::prelude::*;

#[restate_sdk::object]
pub(crate) trait TicketObject {
    async fn reserve() -> Result<bool, HandlerError>;
    async fn unreserve() -> Result<(), HandlerError>;
    #[name = "markAsSold"]
    async fn mark_as_sold() -> Result<(), HandlerError>;
}

pub struct TicketObjectImpl;

impl TicketObject for TicketObjectImpl {
    async fn reserve(&self, ctx: ObjectContext<'_>) -> Result<bool, HandlerError> {
        let status: TicketStatus = ctx
            .get::<Json<TicketStatus>>("status")
            .await?
            .unwrap_or(Json(TicketStatus::Available))
            .into_inner();

        match status {
            TicketStatus::Available => {
                ctx.set("status", Json(TicketStatus::Reserved));
                Ok(true)
            }
            TicketStatus::Reserved | TicketStatus::Sold => Ok(false),
        }
    }

    async fn unreserve(&self, ctx: ObjectContext<'_>) -> Result<(), HandlerError> {
        let status: TicketStatus = ctx
            .get::<Json<TicketStatus>>("status")
            .await?
            .unwrap_or(Json(TicketStatus::Available))
            .into_inner();

        if let TicketStatus::Reserved = status {
            ctx.clear("status");
        }
        Ok(())
    }

    async fn mark_as_sold(&self, ctx: ObjectContext<'_>) -> Result<(), HandlerError> {
        let status: TicketStatus = ctx
            .get::<Json<TicketStatus>>("status")
            .await?
            .unwrap_or(Json(TicketStatus::Available))
            .into_inner();

        if let TicketStatus::Reserved = status {
            ctx.set("status", Json(TicketStatus::Sold));
        }
        Ok(())
    }
}